src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -409,10 +409,17 @@ private unsafe void InnerStartOperationReceiveMessageFrom()
             // WSAMsg also contains a single WSABuffer describing a control buffer.
             PinSocketAddressBuffer();
 
-            // Create and pin a WSAMessageBuffer if none already.
+            // Create a WSAMessageBuffer if none exists yet, and ensure the buffer is pinned.
+            Debug.Assert(
+                !_wsaMessageBufferGCHandle.IsAllocated ||
+                (_wsaMessageBufferGCHandle.Target == _wsaMessageBuffer && _wsaMessageBuffer != null));
+            Debug.Assert((_ptrWSAMessageBuffer != IntPtr.Zero) == _wsaMessageBufferGCHandle.IsAllocated);
```

Comment:
Seems like there should also be an assert that if _ptrWSAMessageBuffer is not null, then it's == Marshal.UnsafeAddrOfPinnedArrayElement(_wsaMessageBuffer, 0).  That's correct, right?

BTW, at some point it would be nice to clean some of this stuff up, the pinning logic is unnecessarily complicated.  A couple thoughts:

(1) Marshal.UnsafeAddrOfPinnedArrayElement(_wsaMessageBuffer, 0) can't be that expensive; rather than storing the raw ptr, we should just do this on demand.
(2) Even better, we could just allocate the WSAMessageBuffer (and WSABuffer also) on the stack when we do the actual native socket call.  Again, I really doubt that the cost of doing this would matter at all.

Thoughts?
 